Optionally use ethtool statistics counters instead of /proc/net/dev

bmork opened the pull request:
RFC really.  Not an actual request to pull...

I am currently experimenting with OpenWrt on Realtek RTL8380 based switches, using a DSA driver for the switch core.  This means that there is a netdev (DSA slave) per switchport with counters in  /proc/net/dev.  But these only count the traffic going between the associated switchport and the cpu port. Which is how DSA is designed AFAICS.  Monitoring that traffic is not particularily interesting. Most packets will be switched to another port and never hit the CPU at all.  Luckily the driver does provide access to the switch counters too, using ethtool.  An actual example:
```
root@gs1900-10hp:/etc# ethtool -S lan10
NIC statistics:
     tx_packets: 28955
     tx_bytes: 3716139
     rx_packets: 1394131
     rx_bytes: 91948697
     ifInOctets: 9353674578
     ifOutOctets: 5205423489
     dot1dTpPortInDiscards: 182569
     ifInUcastPkts: 6989646
     ifInMulticastPkts: 1260808
     ifInBroadcastPkts: 297611
     ifOutUcastPkts: 3873418
     ifOutMulticastPkts: 11175
     ifOutBroadcastPkts: 290
     ifOutDiscards: 0
     .3SingleCollisionFrames: 0
     .3MultipleCollisionFrames: 0
     .3DeferredTransmissions: 0
     .3LateCollisions: 0
     .3ExcessiveCollisions: 0
     .3SymbolErrors: 0
     .3ControlInUnknownOpcodes: 0
     .3InPauseFrames: 0
     .3OutPauseFrames: 0
     DropEvents: 0
     tx_BroadcastPkts: 290
     tx_MulticastPkts: 11175
     CRCAlignErrors: 0
     tx_UndersizePkts: 0
     rx_UndersizePkts: 0
     rx_UndersizedropPkts: 0
     tx_OversizePkts: 3209114
     rx_OversizePkts: 5497377
     Fragments: 0
     Jabbers: 0
     Collisions: 0
     tx_Pkts64Octets: 4148
     rx_Pkts64Octets: 486627
     tx_Pkts65to127Octets: 398437
     rx_Pkts65to127Octets: 1920780
     tx_Pkts128to255Octets: 54306
     rx_Pkts128to255Octets: 74592
     tx_Pkts256to511Octets: 12956
     rx_Pkts256to511Octets: 36775
     tx_Pkts512to1023Octets: 31960
     rx_Pkts512to1023Octets: 20312
     tx_Pkts1024to1518Octets: 173962
     rx_StatsPkts1024to1518Octets: 511602
     tx_Pkts1519toMaxOctets: 3209114
     rx_Pkts1519toMaxOctets: 5497377
     rxMacDiscards: 0
```
The first 4 counters are the netdev counters,  The remaining comes from the switch.  Note the mismatch between for example rx_bytes and ifInOctets.

These patches are an attempt to support using ethtool counters for all or some of the ifTable/ifXTable values. 

The ethtool counter names and availability is obviously driver dependent, so the mapping is made configurable. In theory there can be more than one driver in use on a system for different interfaces, so I decided to create a map per interface.  This makes the configuration too complex for the command line, so libconfuse support is required.  Finally, since a switch driver will provide many similar ports, I included wildcard globbing for the ethtool section names so that you only need one section per driver.

Config file example:
```
# mini-snmpd.conf

location       = "Here"
contact        = "me@example.com"
description    = "ZyXEL GS1900-10HP"

# Vendor OID tree
#vendor         = ".1.3.6.1.4.1"

# true/false, or yes/no
authentication = true
community      = "public"

# MIB poll timeout, sec
timeout        = 1

# Disks to monitor, i.e. mount points in UCD-SNMP-MIB::dskTable
#disk-table     = { "/", }

# Interfaces to monitor, currently only for IF-MIB::ifTable
iface-table    = { "lan1", "lan2", "lan3", "lan4","lan7", "lan8", "lan9", "lan10" }

ethtool "lan*" {
        rx_bytes      = ifInOctets
        rx_mc_packets = ifInMulticastPkts
        rx_bc_packets = ifInBroadcastPkts
        rx_packets    = ifInUcastPkts
        #rx_errors     =
        rx_drops      = dot1dTpPortInDiscards
        tx_bytes      = ifOutOctets
        tx_mc_packets = ifOutMulticastPkts
        tx_bc_packets = fOutBroadcastPkts                                                                                                                                                                                                   
        tx_packets    = ifOutUcastPkts
        #tx_errors     =
        tx_drops      = ifOutBroadcastPkts
}
```

This scratches my itch, but I am a little unsure whether it's generic enough to make sense for the rest of the world.  So I am sending this half finished, without any docs or examples, as an RFC. What do you think?  Does it make sense?  Should it have been done in other ways?

troglobit commented:
This is great stuff, thank you for proposing this!  I'm very familiar with DSA devices, having worked with them for 15+ years at $DAYJOB.  The cascade/cpu ports are mostly interesting when debugging system/driver issues.  Using ethtool is a great idea, since that will also take care of counter wraparound and other issues. I think your remapping support in the .conf is great, only glanced over the code but some helpful defaults or a script/probe-function maybe to extract relevant mappings for end-users would be useful.

Other than that there's only some minor nitpicking from my side: indentation issues, lack of braces around deeply indented code, confuse-macros in header files instead of #ifdef in code.  If you like I could do a full code review?

bmork commented:
Thanks a lot for the positive feedback!.  I should be able to fix up most of the issues without more details...  But if you have time, I would appreciate more detailed pointers wrt how you'd like the #ifdeffed code. I was very unsure about this, and still wonder if it really is a good idea to make the config file format depend on an optional feature.

troglobit commented:
Ah, good point, if it's built w/o ethtool support we wouldn't want it to barf on a .conf file with this statement.  I think the best option is to always allow it and give a syslog warning in `ethtool_xlate_cfg()` if `cfg_size(cfg, "ethtool") > 0`.  This could be a stub that's only built when building w/o ethtool support.

bmork commented:
I believe I have fixed up most of these issues now, except for the proposed script to extract mappings.  I am not sure I can do that generic enough to be useful. I added an example section to the existing config file example.

I pushed the fixups as separate patches for review.  These should obviously be squashed into the original commits before merging

troglobit commented:
Sorry for the late reply, tons of other projects/work/family getting in between!  Looks good, merging :)
